File: nippon/react-native-indicai/ios/IntentEngine.swift
import Foundation
import onnxruntime_objc

private let intentTAG    = "IndicAI.Intent"
private let INTENT_MAX_LEN  = 64
private let INTENT_CLS_ID: Int64 = 101
private let INTENT_SEP_ID: Int64 = 102
private let INTENT_UNK_ID: Int64 = 100
private let INTENT_PAD_ID: Int64 = 0
private let INTENT_EMB_DIM  = 384

/**
 * On-device intent classification using MiniLM-L6 semantic similarity.
 *
 * Mirrors Android IntentEngine.kt exactly:
 *   1. WordPiece-tokenize the English query
 *   2. Run ONNX encoder → L2-normalised 384-dim sentence embedding
 *   3. Cosine similarity against pre-computed intent bank (intent_bank.npy)
 *   4. Nearest neighbour → intent label + similarity score as confidence
 *
 * Files loaded from the extracted intent zip:
 *   intent_encoder.onnx       — MiniLM encoder (output: "sentence_embedding")
 *   tokenizer/vocab.txt       — BERT WordPiece vocabulary
 *   intent_bank.npy           — (N × 384) float32 row-major L2-normalised embeddings
 *   intent_bank_labels.json   — JSON array of N label strings
 */
class IntentEngine {
    private var session: ORTSession?
    private var intentEnv: ORTEnv?
    private var vocab: [String: Int] = [:]
    private var bankVecs: [[Float]] = []     // [N, 384]
    private var bankLabels: [String] = []

    func isLoaded() -> Bool { session != nil }

    func load(modelFile: URL, vocabFile: URL, bankNpyFile: URL, bankLabelsFile: URL) throws {
        let env = try ORTEnv(loggingLevel: .warning)
        self.intentEnv = env
        let opts = try ORTSessionOptions()
        try opts.setIntraOpNumThreads(2)
        try opts.setGraphOptimizationLevel(.all)
        session = try ORTSession(env: env, modelPath: modelFile.path, sessionOptions: opts)
        vocab = try buildVocab(from: vocabFile)
        bankVecs = try loadNpy(from: bankNpyFile)
        bankLabels = try parseLabels(from: bankLabelsFile)
        NSLog("[%@] Loaded: %d bank phrases, %d unique intents",
              intentTAG, bankLabels.count, Set(bankLabels).count)
    }

    func classify(_ text: String) throws -> IntentResult {
        guard let session = session, let env = intentEnv else {
            throw IndicAIError.notInitialized
        }

        // 1. WordPiece tokenize
        let tokenIds = wordpiece(text.lowercased().trimmingCharacters(in: .whitespaces))
        let attn: [Int64] = tokenIds.map { $0 != INTENT_PAD_ID ? 1 : 0 }

        let inputIdsData = tokenIds.withUnsafeBufferPointer { Data(buffer: $0) }
        let attnData     = attn.withUnsafeBufferPointer { Data(buffer: $0) }

        let inputIds = try ORTValue(
            tensorData: NSMutableData(data: inputIdsData),
            elementType: .int64,
            shape: [1, NSNumber(value: INTENT_MAX_LEN)]
        )
        let attnMask = try ORTValue(
            tensorData: NSMutableData(data: attnData),
            elementType: .int64,
            shape: [1, NSNumber(value: INTENT_MAX_LEN)]
        )

        // 2. Encoder → L2-normalised sentence embedding (1, 384)
        let outputs = try session.run(
            withInputs: ["input_ids": inputIds, "attention_mask": attnMask],
            outputNames: ["sentence_embedding"],
            runOptions: nil
        )

        let embData = try outputs["sentence_embedding"]!.tensorData() as Data
        let count = embData.count / MemoryLayout<Float>.size
        let queryVec: [Float] = embData.withUnsafeBytes {
            Array($0.bindMemory(to: Float.self).prefix(count))
        }

        // 3. Cosine similarity — vecs are L2-normalised → dot product = cosine sim
        var bestIdx = 0
        var bestSim: Float = -1
        for (i, row) in bankVecs.enumerated() {
            var dot: Float = 0
            for d in 0..<INTENT_EMB_DIM { dot += row[d] * queryVec[d] }
            if dot > bestSim { bestSim = dot; bestIdx = i }
        }

        let intent = bestIdx < bankLabels.count ? bankLabels[bestIdx] : "unknown"
        NSLog("[%@] classify('%@') → %@ (sim=%.3f)", intentTAG, text, intent, bestSim)
        return IntentResult(intent: intent, confidence: bestSim)
    }

    func close() { session = nil; intentEnv = nil }

    // MARK: - WordPiece tokenizer (matches BERT bert-base-uncased BasicTokenizer)

    private func wordpiece(_ text: String) -> [Int64] {
        // NFD normalize, strip non-spacing marks (diacritics)
        let nfd = text.decomposedStringWithCompatibilityMapping
        var normalized = ""
        for scalar in nfd.unicodeScalars {
            if scalar.properties.generalCategory != .nonspacingMark {
                normalized.unicodeScalars.append(scalar)
            }
        }

        let words = normalized.components(separatedBy: .whitespaces).filter { !$0.isEmpty }
        var pieces: [Int64] = [INTENT_CLS_ID]

        outer: for word in words {
            if pieces.count >= INTENT_MAX_LEN - 1 { break }
            var subTokens: [Int64] = []
            var start = word.startIndex
            var unk = false

            while start < word.endIndex {
                var end = word.endIndex
                var found: Int64? = nil

                while start < end {
                    let sub = start == word.startIndex
                        ? String(word[start..<end])
                        : "##\(String(word[start..<end]))"
                    if let id = vocab[sub] {
                        found = Int64(id)
                        break
                    }
                    end = word.index(before: end)
                }

                if found == nil { unk = true; break }
                subTokens.append(found!)
                start = end
            }

            if unk {
                if pieces.count < INTENT_MAX_LEN - 1 { pieces.append(INTENT_UNK_ID) }
            } else {
                for tok in subTokens {
                    if pieces.count >= INTENT_MAX_LEN - 1 { break outer }
                    pieces.append(tok)
                }
            }
        }
        pieces.append(INTENT_SEP_ID)

        return (0..<INTENT_MAX_LEN).map { $0 < pieces.count ? pieces[$0] : INTENT_PAD_ID }
    }

    // MARK: - Vocab, NPY, Labels loaders

    private func buildVocab(from file: URL) throws -> [String: Int] {
        let text = try String(contentsOf: file, encoding: .utf8)
        var map: [String: Int] = [:]
        let lines = text.components(separatedBy: "\n")
        for (idx, line) in lines.enumerated() {
            let token = line.trimmingCharacters(in: .newlines)
            if !token.isEmpty { map[token] = idx }
        }
        return map
    }

    /// Parse NPY v1.0 file: magic(6) + major(1) + minor(1) + header_len(2 LE) + header + float32 data.
    private func loadNpy(from file: URL) throws -> [[Float]] {
        let bytes = try Data(contentsOf: file)
        guard bytes.count >= 10 else {
            throw IndicAIError.inferenceError("NPY file too small")
        }
        let headerLen = Int(bytes[8]) | (Int(bytes[9]) << 8)
        let offset = 10 + headerLen
        let totalFloats = (bytes.count - offset) / 4
        let rows = totalFloats / INTENT_EMB_DIM
        guard rows > 0 else {
            throw IndicAIError.inferenceError("Empty intent bank NPY")
        }

        var result = [[Float]](repeating: [Float](repeating: 0, count: INTENT_EMB_DIM), count: rows)
        bytes.withUnsafeBytes { raw in
            // All Apple Silicon and x86 iOS devices are little-endian — float32 maps directly
            raw.baseAddress!.advanced(by: offset)
                .withMemoryRebound(to: Float.self, capacity: totalFloats) { floats in
                    for r in 0..<rows {
                        for d in 0..<INTENT_EMB_DIM {
                            result[r][d] = floats[r * INTENT_EMB_DIM + d]
                        }
                    }
                }
        }
        return result
    }

    private func parseLabels(from file: URL) throws -> [String] {
        let data = try Data(contentsOf: file)
        guard let arr = try JSONSerialization.jsonObject(with: data) as? [String] else {
            throw IndicAIError.inferenceError("Invalid intent_bank_labels.json")
        }
        return arr
    }
}

// Shared result type (also used by IndicAIModule)
struct IntentResult {
    let intent: String
    let confidence: Float
}

File: nippon/react-native-indicai/android/src/main/java/com/indicai/sdk/IntentEngine.kt
package com.indicai.sdk

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import android.util.Log
import org.json.JSONArray
import java.io.File
import java.nio.FloatBuffer
import java.nio.LongBuffer
import java.text.Normalizer

private const val TAG      = "IndicAI.Intent"
private const val MAX_LEN  = 64
private const val CLS_ID   = 101L
private const val SEP_ID   = 102L
private const val UNK_ID   = 100L
private const val PAD_ID   = 0L
private const val EMB_DIM  = 384

data class IntentResult(
    val intent: String,
    val confidence: Float,
)

/**
 * On-device intent classification using MiniLM-L6 semantic similarity.
 *
 * How it works (mirrors sentence-transformers on-device):
 *   1. WordPiece-tokenize the English query
 *   2. Run ONNX encoder → normalised 384-dim sentence embedding
 *   3. Cosine similarity against pre-computed intent bank (intent_bank.npy)
 *   4. Nearest neighbour → intent label + similarity score as confidence
 *
 * Files loaded from the intent zip (same S3 / ModelManager flow as other models):
 *   intent_encoder.onnx       — MiniLM encoder (outputs L2-normalised embedding)
 *   tokenizer/vocab.txt       — BERT WordPiece vocabulary
 *   intent_bank.npy           — pre-computed bank embeddings (N × 384 float32, row-major)
 *   intent_bank_labels.json   — JSON array of N label strings matching bank rows
 */
internal class IntentEngine {
    private val env: OrtEnvironment = OrtEnvironment.getEnvironment()
    private var session: OrtSession? = null
    private var vocab: Map<String, Int> = emptyMap()
    private var bankVecs: Array<FloatArray> = emptyArray()   // [N, 384]
    private var bankLabels: List<String> = emptyList()

    fun isLoaded() = session != null

    fun load(modelFile: File, vocabFile: File, bankNpyFile: File, bankLabelsFile: File) {
        session?.close()
        session = env.createSession(
            modelFile.absolutePath,
            OrtSession.SessionOptions().apply {
                setIntraOpNumThreads(2)
                setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
            }
        )
        vocab       = buildVocab(vocabFile)
        bankVecs    = loadNpy(bankNpyFile)
        bankLabels  = parseLabels(bankLabelsFile)
        Log.d(TAG, "Loaded intent engine: ${bankLabels.size} bank phrases, ${bankLabels.toSet().size} intents")
    }

    fun classify(text: String): IntentResult {
        val sess = session ?: throw IllegalStateException("IntentEngine not loaded")

        // 1. Tokenise
        val tokenIds = wordpiece(text.lowercase().trim())
        val attn = LongArray(MAX_LEN) { if (tokenIds[it] != PAD_ID) 1L else 0L }

        val inputIds = OnnxTensor.createTensor(env, LongBuffer.wrap(tokenIds), longArrayOf(1, MAX_LEN.toLong()))
        val attnMask = OnnxTensor.createTensor(env, LongBuffer.wrap(attn),     longArrayOf(1, MAX_LEN.toLong()))

        // 2. Encoder → normalised embedding (1, 384)
        val outputs = sess.run(mapOf("input_ids" to inputIds, "attention_mask" to attnMask))
        val rawValue: Any = outputs[0].value
        Log.d(TAG, "classify rawValue type: ${rawValue.javaClass.name}, bankVecs=${bankVecs.size}")
        @Suppress("UNCHECKED_CAST")
        val queryVec: FloatArray = when (rawValue) {
            is Array<*> -> @Suppress("UNCHECKED_CAST") (rawValue as Array<FloatArray>)[0]
            is FloatArray -> rawValue
            else -> throw IllegalStateException("Unexpected output type: ${rawValue.javaClass.name}")
        }

        // 3. Cosine similarity — vecs are already L2-normalised → dot product = cosine sim
        var bestIdx = 0
        var bestSim = -1f
        for (i in bankVecs.indices) {
            var dot = 0f
            val row = bankVecs[i]
            for (d in 0 until EMB_DIM) dot += row[d] * queryVec[d]
            if (dot > bestSim) { bestSim = dot; bestIdx = i }
        }

        val intent = bankLabels.getOrElse(bestIdx) { "unknown" }
        Log.d(TAG, "classify('$text') → $intent (sim=${String.format("%.3f", bestSim)})")
        return IntentResult(intent, bestSim)
    }

    fun close() {
        session?.close()
        session = null
    }

    // ── WordPiece tokenizer (matches bert-base-uncased BasicTokenizer) ─────────

    private fun wordpiece(text: String): LongArray {
        val normalized = Normalizer.normalize(text, Normalizer.Form.NFD)
            .filter { Character.getType(it) != Character.NON_SPACING_MARK.toInt() }
        val words  = normalized.split(Regex("\\s+")).filter { it.isNotEmpty() }
        val pieces = mutableListOf<Long>()
        pieces.add(CLS_ID)

        outer@ for (word in words) {
            if (pieces.size >= MAX_LEN - 1) break
            val subTokens = mutableListOf<Long>()
            var start = 0
            var unk = false
            while (start < word.length) {
                var end = word.length
                var found: Long? = null
                while (start < end) {
                    val sub = if (start == 0) word.substring(start, end)
                              else "##${word.substring(start, end)}"
                    val id = vocab[sub]
                    if (id != null) { found = id.toLong(); break }
                    end--
                }
                if (found == null) { unk = true; break }
                subTokens.add(found); start = end
            }
            if (unk) {
                if (pieces.size < MAX_LEN - 1) pieces.add(UNK_ID)
            } else {
                for (tok in subTokens) {
                    if (pieces.size >= MAX_LEN - 1) break@outer
                    pieces.add(tok)
                }
            }
        }
        pieces.add(SEP_ID)
        return LongArray(MAX_LEN) { if (it < pieces.size) pieces[it] else PAD_ID }
    }

    private fun buildVocab(file: File): Map<String, Int> {
        val map = mutableMapOf<String, Int>()
        file.useLines { lines -> lines.forEachIndexed { idx, line -> map[line.trimEnd()] = idx } }
        return map
    }

    /**
     * Parse .npy file (NPY format v1.0, float32, C-contiguous 2D array).
     * Returns Array<FloatArray> of shape [rows, EMB_DIM].
     */
    private fun loadNpy(file: File): Array<FloatArray> {
        val bytes = file.readBytes()
        // NPY magic: \x93NUMPY (6 bytes) + major + minor + header_len (2 bytes LE) + header
        var offset = 10 + ((bytes[8].toInt() and 0xFF) or ((bytes[9].toInt() and 0xFF) shl 8))
        val totalFloats = (bytes.size - offset) / 4
        val rows = totalFloats / EMB_DIM
        val result = Array(rows) { FloatArray(EMB_DIM) }
        for (r in 0 until rows) {
            for (d in 0 until EMB_DIM) {
                val i = offset + (r * EMB_DIM + d) * 4
                val bits = ((bytes[i].toInt() and 0xFF))        or
                           ((bytes[i+1].toInt() and 0xFF) shl 8)  or
                           ((bytes[i+2].toInt() and 0xFF) shl 16) or
                           ((bytes[i+3].toInt() and 0xFF) shl 24)
                result[r][d] = java.lang.Float.intBitsToFloat(bits)
            }
        }
        return result
    }

    private fun parseLabels(file: File): List<String> {
        val arr = JSONArray(file.readText())
        return List(arr.length()) { arr.getString(it) }
    }
}
